src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -0,0 +1,81 @@
+    """
+
+    # Get the coding configurations for age and age_category analysis datasets
+    age_category_cc = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.age_category_config is not None:
+                age_category_cc = coding_config
+
+    age_cc = None
+    age_engagement_db_datasets = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.analysis_dataset == age_category_cc.age_category_config.age_analysis_dataset:
+                age_cc = coding_config
+                age_engagement_db_datasets = analysis_dataset_config.engagement_db_datasets
+
+    # Check and impute age_category in age messages only
+    log.info(f"Imputing {age_category_cc.analysis_dataset} for {age_cc.analysis_dataset} messages...")
+    updated_messages_traced_data = []
+    for message in messages_traced_data:
+        if message["dataset"] in age_engagement_db_datasets:
+            age_label = message["labels"][0]
+            age_code = age_cc.code_scheme.get_code_with_code_id(age_label["CodeID"])
+
+            # Impute age_category for this age_code
+            if age_code.code_type == CodeTypes.NORMAL:
+                age_category = None
+                for age_range, category in age_category_cc.age_categories.items():
+                    if age_range[0] <= age_code.numeric_value <= age_range[1]:
+                        age_category = category
+                assert age_category is not None
+                age_category_code = age_category_cc.code_scheme.get_code_with_match_value(age_category)
+            elif age_code.code_type == CodeTypes.META:
+                age_category_code = age_category_cc.code_scheme.get_code_with_meta_code(age_code.meta_code)
+            else:
+                assert age_code.code_type == CodeTypes.CONTROL
+                age_category_code = age_category_cc.code_scheme.get_code_with_control_code(
+                    age_code.control_code)
+
+            age_category_label = CleaningUtils.make_label_from_cleaner_code(
+                age_category_cc.code_scheme, age_category_code, Metadata.get_call_location()
+            )
+
+            # Append this age_category_label to the list of labels for this message, and write-back to TracedData.
+            message_labels = message["labels"].copy()
+            message_labels.insert(0, age_category_label.to_dict())
+            message.append_data(
+                {"labels": message_labels},
+                Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string())
+            )
+
+        updated_messages_traced_data.append(message)
+
+        return updated_messages_traced_data
+
+def run_code_computation_functions(user, messages_traced_data, analysis_dataset_config):
```

Comment:
```suggestion
def impute_codes_by_message(user, messages_traced_data, analysis_dataset_config):
```
or similar - basically, we'll need to distinguish imputations that run message by message and imputations that run by participant

Reply:
Updated thanks.